### PR TITLE
[Widget enhancement] 08 - Forward CoW events up the iFrame

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

Part of the series https://github.com/cowprotocol/cowswap/pull/3812

This PR creates a new updater: `CowEventsUpdater`

This updater responsibility is to subscribe to all events from the `eventEmitter` and for every message, it will forward it up to the iFrame.

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      }
+    })
+
+    allHandlers.forEach((listener) => cowEventEmitter.on(listener as CowEventListener<CowEvents>))
+
+    // Cleanup: Remove all listeners
+    return () => {
+      allHandlers.forEach((listener) => cowEventEmitter.off(listener as CowEventListener<CowEvents>))
+    }
+  }, [])
```
[the remark this answers is not in the document]
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts:32]: fixed c2c0c3a22621d7233f9d42cd93087bc65de65820

File: apps/cowswap-frontend/src/cowEventEmitter.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import { CowEventEmitterImpl } from '@cowprotocol/events'
+
+export const cowEventEmitter = new CowEventEmitterImpl()
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/cowEventEmitter.ts:3]: Cool! Now we have it as a global const, do we still need `CowEventEmitterContext`?

File: apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts (hunk 1 above)
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts:32]: Lint error here.
If you use `cowEventEmitter` as a global const the error will disappear

File: apps/cowswap-frontend/src/cowEventEmitter.ts (hunk 2 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/cowEventEmitter.ts:3]: Hehe, technically not. I just thought its nicer to depend in react on the context for emitting in components. The context exposes the interface only, not the implementation. 

TBH, I think is not great the global var, I was not planning to do it, but I also hated the idea of adding it in the redux action payload. So you would prefer to kill the context?

File: apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts (hunk 1 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts:32]: The lint issue was corrected at some point later https://github.com/cowprotocol/cowswap/blob/a6026f029c1cf5795ea0ff55d58fb5804ee0e1ee/apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts#L34

Comment (anxolin) [reply apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts:32]: I thought a bit more about this, you convinced me. I will kill the context, and leave a single eventEmitter. Expect a PR soon!

Comment (anxolin) [reply apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts:32]: https://github.com/cowprotocol/cowswap/pull/3839

File: apps/cowswap-frontend/src/cowEventEmitter.ts (hunk 2 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/cowEventEmitter.ts:3]: https://github.com/cowprotocol/cowswap/pull/3839

**Comment by anxolin** [comment]
Merging without approvals. Let me know if there's anything else to address

**Pull request merged**